### PR TITLE
Introduce qmcplusplus::isnan

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
Introduce custom isnan to make it always work.

## What type(s) of changes does this code introduce?
- New feature

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

**Comment by ye-luo** [comment]
Test this please

**Comment by PDoakORNL** [comment]
So we need to support some platform without `std::isnan`?

**Comment by ye-luo** [comment]
> So we need to support some platform without `std::isnan`?

When `-fffast-math` is enabled, std::isnan always return false and I cannot implement safe-guards. qmcplusplus::isnan remains effective.

**Review by prckent — CHANGES_REQUESTED** [review]
This will be useful to have since unfortunately there are sightings of NaN in optimized builds where std::isnan is not helpful. AFAICT checking a!=a is the preferred way to do this.

Can you please add comments to the changes in math.hpp,.cpp explaining the functions?

**Review by prckent — APPROVED** [review]

**Comment by prckent** [comment]
Test this please

File: src/Platforms/CPU/CMakeLists.txt [hunk 1]
```diff
@@ -2,14 +2,17 @@
...
+#// Copyright (c) 2023 QMCPACK developers.
 #//
 #// File developed by: Ye Luo, yeluo@anl.gov, Argonne National Laboratory
 #//
 #// File created by: Ye Luo, yeluo@anl.gov, Argonne National Laboratory
 #//////////////////////////////////////////////////////////////////////////////////////
 
-add_library(platform_cpu_runtime INTERFACE)
+# To enable a custom qmcplusplus::isnan, remove fast-math. Only the file scope CMAKE_CXX_FLAGS gets affected.
+string(REPLACE " -ffast-math" "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
```
Comment (markdewing) [inline src/Platforms/CPU/CMakeLists.txt:13]: Do all compilers use a flag with the same name to perform the optimization that affects isnan?

Comment (ye-luo) [reply src/Platforms/CPU/CMakeLists.txt:13]: All the compilers that have `-ffast-math` behave the same. I explicitly tested nvhpc which doesn't support `-ffast-math` and passes the added tests.

Comment (prckent) [reply src/Platforms/CPU/CMakeLists.txt:13]: In any case this PR still expands the range of cases for which we can catch a NaN. Hopefully the code path won't get much use...

**Pull request merged**